File: src/camera/Camera.cpp
#include "camera/Camera.h"
#include "core/Input.h"

#include <glm/gtc/matrix_transform.hpp>

Camera::Camera(float aspect)
{
    position = {0.0f, 0.0f, 5.0f};

    this->aspect = aspect;

    fov = 45.0f;
    nearPlane = 0.1f;
    farPlane = 100.0f;
    speed = 2.5f;

    yaw = -90.0f;
    pitch = 0.0f;

    sensitivity = 0.1f;

    firstMouse = true;
}

glm::mat4 Camera::getViewMatrix()
{
    glm::vec3 front;

    front.x =
        cos(glm::radians(yaw)) *
        cos(glm::radians(pitch));

    front.y =
        sin(glm::radians(pitch));

    front.z =
        sin(glm::radians(yaw)) *
        cos(glm::radians(pitch));

    glm::vec3 dir = glm::normalize(front);

    return glm::lookAt(
        position,
        position + dir,
        glm::vec3(0,1,0)
    );
}

glm::mat4 Camera::getProjectionMatrix()
{
    return glm::perspective(
        glm::radians(fov),
        aspect,
        nearPlane,
        farPlane
    );
}

void Camera::update(float dt)
{
    glm::vec3 front = getFront();

    glm::vec3 right =
        glm::normalize(
            glm::cross(
                front,
                glm::vec3(0,1,0)
            )
        );

    if (Input::isKeyPressed(GLFW_KEY_W))
        position += front * speed * dt;

    if (Input::isKeyPressed(GLFW_KEY_S))
        position -= front * speed * dt;

    if (Input::isKeyPressed(GLFW_KEY_A))
        position -= right * speed * dt;

    if (Input::isKeyPressed(GLFW_KEY_D))
        position += right * speed * dt;


    double x = Input::getMouseX();
    double y = Input::getMouseY();

    if (firstMouse)
    {
        lastX = x;
        lastY = y;
        firstMouse = false;
    }

    double dx = x - lastX;
    double dy = lastY - y;

    lastX = x;
    lastY = y;

    yaw += dx * sensitivity;
    pitch += dy * sensitivity;

    if (pitch > 89.0f) pitch = 89.0f;
    if (pitch < -89.0f) pitch = -89.0f;
}

glm::vec3 Camera::getFront()
{
    glm::vec3 front;

    front.x =
        cos(glm::radians(yaw)) *
        cos(glm::radians(pitch));

    front.y =
        sin(glm::radians(pitch));

    front.z =
        sin(glm::radians(yaw)) *
        cos(glm::radians(pitch));

    return glm::normalize(front);
}

void Camera::resetMouse()
{
    firstMouse = true;
}

File: src/camera.rs
use glam::{Mat4, Vec3};
use glfw::{Action, Key, Window};

pub struct Camera {
    pub position: Vec3,
    pub fov: f32,
    pub near_plane: f32,
    pub far_plane: f32,
    pub yaw: f32,
    pub pitch: f32,
    aspect: f32,
}

impl Camera {
    pub fn new(aspect: f32) -> Self {
        Self {
            position: Vec3::new(0.0, 0.0, 5.0),
            fov: 45.0_f32.to_radians(),
            near_plane: 0.1,
            far_plane: 100.0,
            yaw: -90.0,
            pitch: 0.0,
            aspect,
        }
    }

    pub fn set_aspect(&mut self, aspect: f32) {
        self.aspect = aspect;
    }

    pub fn front(&self) -> Vec3 {
        let yaw = self.yaw.to_radians();
        let pitch = self.pitch.to_radians();
        Vec3::new(
            yaw.cos() * pitch.cos(),
            pitch.sin(),
            yaw.sin() * pitch.cos(),
        )
        .normalize()
    }

    pub fn view_matrix(&self) -> Mat4 {
        let dir = self.front();
        Mat4::look_at_rh(self.position, self.position + dir, Vec3::Y)
    }

    pub fn projection_matrix(&self) -> Mat4 {
        Mat4::perspective_rh_gl(self.fov, self.aspect, self.near_plane, self.far_plane)
    }
}

/// First-person fly camera (WASD, QE, mouse look).
pub struct FlyCamera {
    move_speed: f32,
    sensitivity: f32,
    enabled: bool,
    first_mouse: bool,
    last_x: f64,
    last_y: f64,
}

impl Default for FlyCamera {
    fn default() -> Self {
        Self {
            move_speed: 2.5,
            sensitivity: 0.1,
            enabled: false,
            first_mouse: true,
            last_x: 0.0,
            last_y: 0.0,
        }
    }
}

impl FlyCamera {
    pub fn set_enabled(&mut self, enabled: bool) {
        self.enabled = enabled;
    }

    pub fn reset_mouse(&mut self) {
        self.first_mouse = true;
    }

    pub fn update(&mut self, camera: &mut Camera, window: &Window, dt: f32) {
        if !self.enabled {
            return;
        }

        let fast = window.get_key(Key::LeftShift) == Action::Press
            || window.get_key(Key::RightShift) == Action::Press;
        let speed = self.move_speed * if fast { 3.0 } else { 1.0 };

        let front = camera.front();
        let right = front.cross(Vec3::Y).normalize();

        if window.get_key(Key::W) == Action::Press {
            camera.position += front * speed * dt;
        }
        if window.get_key(Key::S) == Action::Press {
            camera.position -= front * speed * dt;
        }
        if window.get_key(Key::A) == Action::Press {
            camera.position -= right * speed * dt;
        }
        if window.get_key(Key::D) == Action::Press {
            camera.position += right * speed * dt;
        }
        if window.get_key(Key::E) == Action::Press {
            camera.position += Vec3::Y * speed * dt;
        }
        if window.get_key(Key::Q) == Action::Press {
            camera.position -= Vec3::Y * speed * dt;
        }

        let (x, y) = window.get_cursor_pos();
        if self.first_mouse {
            self.last_x = x;
            self.last_y = y;
            self.first_mouse = false;
        }

        let dx = x - self.last_x;
        let dy = self.last_y - y;
        self.last_x = x;
        self.last_y = y;

        camera.yaw += dx as f32 * self.sensitivity;
        camera.pitch += dy as f32 * self.sensitivity;
        camera.pitch = camera.pitch.clamp(-89.0, 89.0);
    }
}
